csl/core/engine.lua

```diff
@@ -0,0 +1,1010 @@
+      local groupCond = "_group_" .. self.groupState.count
+      self:_addGroupVariable(groupCond, true)
+   end
+   SU.debug("csl", "Leave group", self.groupState.count, "level", #self.groupQueue, suppressGroup and "(suppressed)" or "(rendered)")
+   return rendered
+end
+
+function CslEngine:_addGroupVariable (variable, value)
+   SU.debug("csl", "Group variable", variable, value and "true" or "false")
+   self.groupState.variables[variable] = value and true or false
+end
+
+-- RENDERING ATTRIBUTES (strip-periods, affixes, formatting, text-case, display, quotes, delimiter)
+
+function CslEngine:_punctuation_extra (t, options)
+   if self.extras.localizedPunctuation then
+      -- non-standard: localized punctuation
+      return t:gsub("[,;:]", function (c) return self.punctuation[c] or c end)
+   end
+   return t
+end
+
+function CslEngine:_render_stripPeriods (t, options)
+   if t and options["strip-periods"] and t:sub(-1) == "." then
+      t = t:sub(1, -2)
+   end
+   return t
+end
+
+function CslEngine:_render_affixes (t, options)
+   if not t then
+      return
+   end
+   if options.prefix then
+      local pref = self:_punctuation_extra(options.prefix)
+      t = pref .. t
+   end
+   if options.suffix then
+      local suff = self:_punctuation_extra(options.suffix)
+      t = t .. suff
+   end
+   return t
+end
+
+function CslEngine:_render_formatting (t, options)
+   if not t then
+      return
+   end
+   if options["font-style"] == "italic" then -- FIXME: also normal, oblique, and how nesting is supposed to work?
+      t = "<em>" .. t .. "</em>"
+   end
+   if options["font-variant"] == "small-caps" then
+      t = "<font features=\"+smcp\">" .. t .. "</font>"
+   end
+   if options["font-weight"] == "bold" then -- FIXME: also light, normal, and how nesting is supposed to work?
+      t = "<strong>" .. t .. "</strong>"
+   end
+   if options["text-decoration"] == "underline" then
+      t = "<underline>" .. t .. "</underline>"
+   end
+   if options["vertical-align"] == "sup" then
+      t = "<textsuperscript>" .. t .. "</textsuperscript>"
+   end
+   if options["vertical-align"] == "sub" then
+      t = "<textsubscript>" .. t .. "</textsubscript>"
+   end
+   return t
+end
+
+function CslEngine:_render_textCase (t, options)
+   if not t then
+      return
+   end
+   if options["text-case"] then
+      t = self.locale:case(t, options["text-case"])
+   end
+   return t
+end
+
+function CslEngine:_render_display (t, options)
+   if not t then
+      return
+   end
+   -- if options.display then
+      -- FIXME Add rationale for not supporting it...
+      -- Keep silent: it's not a critical feature yet
+      -- SU.warn("CSL display not implemented")
+   -- end
+   return t
+end
+
+function CslEngine:_render_quotes (t, options)
+   if t and options.quotes then
+      -- Smart transform straight quotes in the input to localized inner quotes.
+      t = luautf8.gsub(t, "([%s%p])\"", "%1" .. self.punctuation.open_inner_quote)
```

Comment:
While this works for inner quotes surrounded by text (e.g. `Some "special" text` --> `Some ‘special’ text`), we also need to take into account quotes at the start or end of the input (e.g. `Some "special text"` --> `Some ‘special text’`)
```
      t = luautf8.gsub(t, "^\"", self.punctuation.open_inner_quote)
      t = luautf8.gsub(t, "\"$", self.punctuation.close_inner_quote)
```

Reply:
Note to myself if I need ever to need remember it: checked on `dragon:swank2017b`

Reply:
Done